src/java.base/share/classes/jdk/internal/io/JdkConsoleImpl.java

```diff
@@ -348,19 +330,38 @@ public int read(char[] cbuf, int offset, int length)
         }
     }
 
-    ConsoleImpl() {
-        readLock = new Object();
-        writeLock = new Object();
-        out = StreamEncoder.forOutputStreamWriter(
-                new FileOutputStream(FileDescriptor.out),
-                writeLock,
-                CHARSET);
-        pw = new PrintWriter(out, true) { public void close() {} };
-        formatter = new Formatter(out);
-        reader = new LineReader(StreamDecoder.forInputStreamReader(
-                new FileInputStream(FileDescriptor.in),
-                readLock,
-                CHARSET));
-        rcb = new char[1024];
+    @Override
+    public JdkConsole console(boolean isTTY, Charset charset) {
+        if (isTTY) {
+            if (INSTANCE == null) {
+                this.charset = charset;
+                readLock = new Object();
+                writeLock = new Object();
+                out = StreamEncoder.forOutputStreamWriter(
+                        new FileOutputStream(FileDescriptor.out),
+                        writeLock,
+                        charset);
+                pw = new PrintWriter(out, true) {
+                    public void close() {
+                    }
+                };
+                formatter = new Formatter(out);
+                reader = new LineReader(StreamDecoder.forInputStreamReader(
+                        new FileInputStream(FileDescriptor.in),
+                        readLock,
+                        charset));
+                rcb = new char[1024];
+                INSTANCE = this;
+            }
+            return INSTANCE;
+        } else {
+            // not instantiable
+            return null;
+        }
+    }
+
+    // For convenience
+    public static JdkConsole getJdkConsole(Charset cs) {
+        return INSTANCE != null ? INSTANCE : new JdkConsoleImpl().console(true, cs);
```

Comment:
Would this singleton/caching be necessary? Or could we expect that the caller of this method would cache it (like the `java.io.Console` actually does right now)?
In theory, this current implementation in this method can potentially return an incorrect (cached) instance on a second call if the `Charset` that's passed is different from the one that was used by the cached instance that is returned. But, since we know/expect this method to be called only by the `java.io.Console` and with the same `Charset` instance, I think this is OK. I do think that if we remove the caching/singleton logic here, this class could be a bit more simpler.